File: src/control_translation/policy_reader/fixtures.py
"""Deterministic fixture PolicyReader.

Used by default (RUN_MODE=fixture) and by all offline tests. Represents a
plausible current policy snapshot for each fixture target technology.
"""

from __future__ import annotations

from control_translation.policy_reader.base import PolicySnapshot


_FIXTURE_SNAPSHOTS: dict[tuple[str, str], PolicySnapshot] = {
    ("akamai-waf", "akamai-policy:example:rev-17"): PolicySnapshot(
        snapshot_id="policy-snapshot:akamai:example:rev-17",
        target_technology="akamai-waf",
        target_policy_context_id="akamai-policy:example:rev-17",
        existing_rule_ids=["rule-1001", "rule-1002"],
        existing_rule_summaries=[
            "rule-1001: block known SQLi patterns in query string",
            "rule-1002: rate-limit login endpoint",
        ],
    ),
    ("firewall-generic", "firewall-policy:example:rev-4"): PolicySnapshot(
        snapshot_id="policy-snapshot:firewall:example:rev-4",
        target_technology="firewall-generic",
        target_policy_context_id="firewall-policy:example:rev-4",
        existing_rule_ids=["fw-rule-10", "fw-rule-11"],
        existing_rule_summaries=[
            "fw-rule-10: deny inbound on legacy management port",
            "fw-rule-11: allow outbound DNS",
        ],
    ),
}


class FixturePolicyReader:
    """Deterministic offline PolicyReader implementation."""

    def read_snapshot(
        self, target_technology: str, target_policy_context_id: str
    ) -> PolicySnapshot | None:
        return _FIXTURE_SNAPSHOTS.get((target_technology, target_policy_context_id))

File: go-api/internal/policy/fixtures.go
// Package policy reads the current live policy for a target technology.
//
// No real Akamai API client exists yet; only the deterministic fixture reader
// is provided. A real reader is future work: until it exists, the conflict
// gate compares candidates against bundled snapshots, not live policy.
package policy

import "github.com/ATT-CSO/control-translation/go-api/internal/adapters"

// Reader is what the capability needs from the world to read current policy.
type Reader interface {
	// ReadSnapshot returns the current snapshot, or nil when unavailable.
	ReadSnapshot(targetTechnology, targetPolicyContextID string) *adapters.PolicySnapshot
}

type key struct{ technology, context string }

var fixtureSnapshots = map[key]adapters.PolicySnapshot{
	{"akamai-waf", "akamai-policy:example:rev-17"}: {
		SnapshotID:            "policy-snapshot:akamai:example:rev-17",
		TargetTechnology:      "akamai-waf",
		TargetPolicyContextID: "akamai-policy:example:rev-17",
		ExistingRuleIDs:       []string{"rule-1001", "rule-1002"},
		ExistingRuleSummaries: []string{
			"rule-1001: block known SQLi patterns in query string",
			"rule-1002: rate-limit login endpoint",
		},
		IsFixture: true,
	},
}

// FixtureReader is the deterministic offline Reader implementation.
type FixtureReader struct{}

// ReadSnapshot implements Reader.
func (FixtureReader) ReadSnapshot(targetTechnology, targetPolicyContextID string) *adapters.PolicySnapshot {
	snapshot, ok := fixtureSnapshots[key{targetTechnology, targetPolicyContextID}]
	if !ok {
		return nil
	}
	return &snapshot
}
